xaitk_saliency/utils/tools/drise_from_config.py

```diff
@@ -0,0 +1,130 @@
+        """
+        This helper function takes detections as returned by an implementation
+        of DetectImageObjects and returns a matrix formatted for use with the
+        DRISE algorithm, as returned by the format_detection() function.
+        """
+
+        if len(dets) == 0:
+            return np.asarray([])
+
+        bbox_mat = np.asarray([
+            [
+                bbox.min_vertex[0],
+                bbox.min_vertex[1],
+                bbox.max_vertex[0],
+                bbox.max_vertex[1]
+            ] for bbox in [det[0] for det in dets]
+        ])
+
+        class_mat = np.asarray([
+            list(class_dict.values())
+            for class_dict in [det[1] for det in dets]
+        ])
+
+        return format_detection(bbox_mat, class_mat)
+
+    # DetectImageObjects implementation
+    blackbox_detector = from_config_dict(config_dict, DetectImageObjects.get_impls())
+
+    img_perturber = RISEGrid(num_perturbations, 8, 0.5, seed=0, threads=4)
+    sal_generator = DRISEScoring()
+
+    # List of saliency maps for all images
+    sal_maps_set = []
+
+    for i, ref_img in enumerate(ref_imgs):
+
+        if verbose:
+            print(f'[{i+1}/{len(ref_imgs)}] Getting reference detections.')
+
+        ref_dets = iter(blackbox_detector([ref_img])).__next__()
+        num_dets = sum(1 for det in ref_dets)
+
+        if verbose:
+            print(f'[{i+1}/{len(ref_imgs)}] {num_dets} reference detections found.')
+
+        # Return empty list if no reference detections are made
+        if num_dets == 0:
+            sal_maps_set.append(np.asarray([]).reshape(0, ref_img.shape[0], ref_img.shape[1]))
+            continue
+
+        if verbose:
+            print(f'[{i+1}/{len(ref_imgs)}] Perturbing image.')
+
+        pert_masks = img_perturber(ref_img)
+        pert_imgs = occlude_image_batch(ref_img, pert_masks)
+
+        if verbose:
+            print(f'[{i+1}/{len(ref_imgs)}] Getting perturbation detections.')
+
+        # Detections on all perturbed images
+        pert_dets_set = blackbox_detector(pert_imgs)
+
+        if verbose:
+            print(f'[{i+1}/{len(ref_imgs)}] Generating saliency maps.')
+
+        ref_dets_mat = dets_to_formatted_mat(ref_dets)
+
+        pert_dets_arrays = [dets_to_formatted_mat(pert_dets) for pert_dets in pert_dets_set]
+
+        # Use max detections found for a perturbed image as the number of proposals
+        num_props = max([array.shape[0] for array in pert_dets_arrays])
+
+        # Row of random numbers and negative objectiveness to pad proposals with
+        pad_prop = np.random.rand(ref_dets_mat.shape[1])
+        pad_prop[4] = -1
+
+        for i, dets_array in enumerate(pert_dets_arrays):
+            num_pert_dets = dets_array.shape[0]
+
+            if num_pert_dets < num_props:
+                pert_dets_arrays[i] = np.vstack((
+                    dets_array.reshape(dets_array.shape[0], ref_dets_mat.shape[1]),
```

Comment:
Why reshape? I would have thought that the matrices that come out of `dets_to_formatted_mat` would be have the same number of columns.

Reply:
I reshape here for anytime when `dets_array` is an empty and doesn't have a second dimension. Is there a better way to do this?